### PR TITLE
palomar (search) iteration

### PR DESCRIPTION
**Pull request by bnewbold**
Larger refactors in this branch:

- [x] local docker dev env documented
- [x] specify mappings (schemas) for post and profile indices
- [x] transform raw records in to the index schemas
- [x] different doc _id syntax
- [x] skip read+deserialization of records other than profile and post, for efficiency
- [x] don't store records in database; database only used for firehose cursor state
- [x] switch to informal /xrpc/app.bsky.unspecced.search*Skeleton endpoints
- [x] return only skeleton responses (eg, AT-URI or DID lists)
- [x] handle non-success OpenSearch responses as errors
- [x] auto-create indices with schema when in indexing mode (not READONLY) (with `go:embed` schemas)
- [x] switch logging to `log/slog`, including echo integration
- [x] use `atproto/identity` package for identity caching and handling, not `User` database record
- [x] merged in backfill worker code
- [x] use `analysis-icu` plugin for (hopefully) better internationalized search
- [x] special typeahead indexing and query parameter
- [x] basic/simple query string parsing, which should be safe, supports quoted phrases, and `from:` filtering

This branch includes a couple small commits to SDK code, which i've cherry-picked out as separate PRs for easier review.

See also Lexicon PR in atproto repo: https://github.com/bluesky-social/atproto/pull/1594

This is not compatible with the previous version of `palomar` at the HTTP API, opensearch index, or database schema levels. The config vars should be backwards compatible. The operational plan for staging and prod is to deploy this as an entirely new environment (eg, "prod2", "staging2"), get everything backfilled, and then flip over the AppView and then client app to use the lexicons/endpoints instead of the older version.

----

I think this is ready for review, merge, and deploy to staging. Some things to check before prod:

- [ ] compare index size and performance to existing version/schema
- [ ] real-world testing of profile typeahead (eg, do we need fuzzy?)
- [ ] real-world search relevancy checks
- [ ] real-world CJK text analysis checks (https://github.com/bluesky-social/indigo/issues/302)

Out of scope for this PR:

- [ ] deal with `created_at` timestamp not being reliable, by adding a `sort_at` hybrid field, for future "sort by date"
- [ ] instrumentation and metrics (Jaz to implement on top of this branch)
- [ ] better bulk indexing performance, especially during backfill: disable refresh during backfill? longer refresh window? bulk (batch) indexing would be best
- [x] integrate a better identity service/cache; current is probably Ok in context of backfill. or perhaps just bump the cache size to ~50k or ~100k identities in prod?

### CODE REVIEW

File: cmd/palomar/main.go
Comment (ericvolp12) [inline cmd/palomar/main.go:149]: We could probably bump this to like 200k or smth considering we've got 1M users now

File: search/firehose.go
Comment (ericvolp12) [inline search/firehose.go:63]: A `cursor` of 0 asks for the farthest back event that the host is willing to provide. Maybe we want to default the value to `-1` as s sentinel so we can allow for an explicit 0 cursor?

Doesn't need to be blocking but just a thing to think about.

File: search/handlers.go
Comment (ericvolp12) [inline search/handlers.go:29]: Is this a cursor or an offset?

File: search/handlers.go
Comment (ericvolp12) [inline search/handlers.go:136]: Do we want this to fail the entire query or just log it with some context and move to the next result?

File: search/handlers.go
Comment (ericvolp12) [inline search/handlers.go:141]: Same as above

File: search/handlers.go
Comment (ericvolp12) [inline search/handlers.go:151]: Is this supposed to be here?

File: search/handlers.go
Comment (ericvolp12) [inline search/handlers.go:174]: Similarly, do we want to error here or log it and keep going?

File: search/query.go
Comment (ericvolp12) [inline search/query.go:51]: Are these a bit redundant? `offset+size` can't be > `2000` with the explicit size and offset rules, right?

File: search/query.go
Comment (ericvolp12) [inline search/query.go:65]: I wonder how this is going to impact searching for words that people have in their handles

File: search/query.go
Comment (ericvolp12) [inline search/query.go:172]: Why is this in an `if true`?

File: search/server.go
Comment (ericvolp12) [inline search/server.go:89]: Should these be runtime configurable?

**Review by ericvolp12 — APPROVED** [review]
Overall looks solid, have a few comments on how we want to handle different error cases and some dead code bits etc.

File: cmd/palomar/main.go
Comment (bnewbold) [reply cmd/palomar/main.go:149]: A lot fewer accounts will have emitted an event within the TLL window though. Probably should be configurable but i'll set to 200k for now, this is an indexing service.

File: search/firehose.go
Comment (bnewbold) [reply search/firehose.go:63]: We'll probably want to be able to configure this, the same way Kafka lets you default a consumer group to "start at current offset", "start at earliest available", "start at a configured time delta", etc.

For now it is in the "zero means empty/nil" situation. I'm going to leave as-is, which I think is a sane default with the backfill stuff now handling broader enumeration of backfill work.

File: search/handlers.go
Comment (bnewbold) [reply search/handlers.go:136]: This is a pretty bad error which should "never" happen, so I think we want to fail-and-bail on this situation.

An example where this would happen is if there was a config mistake where a query worker got pointed at the wrong elasticsearch index (either post instead of profile, or an older schema version). I think in that kind of situation we want to fail loudly.

File: search/handlers.go
Comment (bnewbold) [reply search/handlers.go:151]: whoops!

File: search/handlers.go
Comment (bnewbold) [reply search/handlers.go:174]: same as above: this is a bad and very unexpected situation

File: search/handlers.go
Comment (bnewbold) [reply search/handlers.go:141]: same as above: this is a bad and very unexpected situation

File: search/handlers.go
Comment (bnewbold) [reply search/handlers.go:179]: same as above

File: search/query.go
Comment (bnewbold) [reply search/query.go:51]: yeah, it is redundant. I guess should decide whether we want this check here or in the other place. I like having it just before query to have confidence it is always applied; but also verifying when HTTP requests are parsed is idiomatic. hrm.

File: search/query.go
Comment (bnewbold) [reply search/query.go:65]: For the post schema, we don't include handles in the document, or in "everything", so should not impact.

File: search/query.go
Comment (ericvolp12) [reply search/query.go:51]: The check is fine you're just checking x+y > 5000 || x > 1000 || y > 1000 which is redundant since x+y will only be > 5000 if one of them is > 1000.

File: search/query.go
Comment (bnewbold) [reply search/query.go:172]: I can't remember why now... this code block is weird/annoying, will clean it up

File: search/server.go
Comment (bnewbold) [reply search/server.go]: yup, added

File: search/query.go
Comment (bnewbold) [reply search/query.go]: oh, hah, yeah. updated with better and no-longer-redundant sanity check:

`offset+size > 10000 || size > 250 || offset > 10000 || offset < 0 || size < 0`

Comment (bnewbold) [reply search/query.go]: (well, I guess it is still slightly redundant, but :shrug:)

**Comment by bnewbold** [comment]
Addressed a bunch of these. The main outstanding thing is failing hard on some of the query error cases. I *think* that is the right thing to do, but could be wrong and we could get bitten.

I'll probably merge this later/tomorrow after thinking a bit more.

File: search/firehose.go
Comment (ericvolp12) [reply search/firehose.go:63]: Just clarifying, this means if you provide no cursor we'll start from as far back as the BGS/PDS will allow us, not from live.

Comment (bnewbold) [reply search/firehose.go:63]: Looking at this again, I guess I had a typo and it should be `RawQuery` not `RawPath`.

But if we don't provide a `cursor` HTTP query param at all, shouldn't the PDS (or BGS) give us the current stream, not start at the beginning? The intent of this snippet of code is to not set the parameter at all, instead of defaulting to `cursor=0` (which would scroll back to oldest available).

Comment (ericvolp12) [reply search/firehose.go]: Oh oops yeah you're right, I read that wrong

**Pull request merged**
